File: 06_strings/06_04_longest_common_prefix/longest_common_prefix.cpp
#include <bits/stdc++.h>

using namespace std;

string longestCommonPrefix(vector<string> &strs)
{
    if (strs.empty())
    {
        return "";
    }

    sort(strs.begin(), strs.end());
    int index = 0;
    int length = strs.size();
    for (int i = 0; i < length; i++)
    {
        if (strs[0][i] == strs.back()[i])
        {
            index++;
        }
        else
        {
            break;
        }
    }
    return strs[0].substr(0, index);
}

int main()
{
    // Test case 1: Basic example
    vector<string> input1 = {"flower", "flow", "flight"};
    string result1 = longestCommonPrefix(input1);
    cout << "Test Case 1: Input: {";
    for (const auto &str : input1)
    {
        cout << "\"" << str << "\", ";
    }
    cout << "}, Result: \"" << result1 << "\"" << endl;

    // Test case 2: No common prefix
    vector<string> input2 = {"dog", "racecar", "car"};
    string result2 = longestCommonPrefix(input2);
    cout << "Test Case 2: Input: {";
    for (const auto &str : input2)
    {
        cout << "\"" << str << "\", ";
    }
    cout << "}, Result: \"" << result2 << "\"" << endl;

    // Test case 3: Common prefix at the beginning
    vector<string> input3 = {"apple", "apricot", "apology"};
    string result3 = longestCommonPrefix(input3);
    cout << "Test Case 3: Input: {";
    for (const auto &str : input3)
    {
        cout << "\"" << str << "\", ";
    }
    cout << "}, Result: \"" << result3 << "\"" << endl;

    // Test case 4: Empty vector
    vector<string> input4 = {};
    string result4 = longestCommonPrefix(input4);
    cout << "Test Case 4: Input: {";
    for (const auto &str : input4)
    {
        cout << "\"" << str << "\", ";
    }
    cout << "}, Result: \"" << result4 << "\"" << endl;

    // Test case 5: Single-word vector
    vector<string> input5 = {"programming"};
    string result5 = longestCommonPrefix(input5);
    cout << "Test Case 5: Input: {";
    for (const auto &str : input5)
    {
        cout << "\"" << str << "\", ";
    }
    cout << "}, Result: \"" << result5 << "\"" << endl;

    return 0;
}

File: 06_strings/06_04_longest_common_prefix/longest_common_prefix.js
const longestCommonPrefix = (strs = []) => {
    if (!strs.length) {
        return '';
    }

    strs.sort();
    let index = 0;
    const length = strs[0].length;
    while (index < length) {
        if (strs[0][index] === strs[strs.length - 1][index]) {
            index++;
        } else {
            break;
        }
    }
    return strs[0].substring(0, index);
}

// Test case 1: Basic example
let input1 = ["flower", "flow", "flight"];
let result1 = longestCommonPrefix(input1);
console.log(`Test Case 1: Input: [${input1}], Result: "${result1}"`);

// Test case 2: No common prefix
let input2 = ["dog", "racecar", "car"];
let result2 = longestCommonPrefix(input2);
console.log(`Test Case 2: Input: [${input2}], Result: "${result2}"`);

// Test case 3: Common prefix at the beginning
let input3 = ["apple", "apricot", "apology"];
let result3 = longestCommonPrefix(input3);
console.log(`Test Case 3: Input: [${input3}], Result: "${result3}"`);

// Test case 4: Empty array
let input4 = [];
let result4 = longestCommonPrefix(input4);
console.log(`Test Case 4: Input: [${input4}], Result: "${result4}"`);

// Test case 5: Single-word array
let input5 = ["programming"];
let result5 = longestCommonPrefix(input5);
console.log(`Test Case 5: Input: [${input5}], Result: "${result5}"`);